### ISSUE TITLE
eviction-family caches: eviction corrupts attention output when triggered (mask="causal" shortcut mismatches query/key length) — confirmed in snapkv, squeeze, streaming_llm

### ISSUE THREAD

**Issue by rajveer43** [rajveer43/VeloxQuant-MLX#370]
## Summary

`SnapKVKVCache`'s prefill eviction (`_process_prefill`) evicts tokens **before** returning keys/values to `mlx_lm`'s attention layer for that same prefill step — but that step's *own* queries still have the full, un-evicted length. The mismatch between query count and (now-smaller) key count silently falls into MLX's `mask="causal"` fast-path, which assumes keys are a contiguous trailing window aligned to the queries — false once eviction has dropped a sparse, non-contiguous subset of positions. The result is severely corrupted prefill attention with no error raised.

Discovered while live-verifying VeloxQuant-Studio issue #27, after fixing the (unrelated, routine) `merge()`/`is_trimmable()` batching-substitution bug: fixing `merge()` was what let snapkv's real eviction path run for the first time in `mlx_lm.server` — previously, `mlx_lm`'s batching code silently substituted a `BatchKVCache` (no eviction at all) for every request, fully masking this bug.

## Repro

```python
import mlx.core as mx
from mlx_lm import load, stream_generate
from veloxquant_mlx.cache.base import KVCacheConfig, KVCacheBuilder

model, tokenizer = load("mlx-community/Qwen2.5-0.5B-Instruct-4bit")
prompt_text = ("Please write a detailed six-paragraph essay about the history and "
    "evolution of computer operating systems, covering early batch systems, "
    "time-sharing, personal computers, and modern mobile and cloud platforms.")
messages = [{"role": "user", "content": prompt_text}]
prompt = tokenizer.apply_chat_template(messages, add_generation_prompt=True)  # 65 tokens

for budget in (200, 60, 50, 40, 35):
    cfg = KVCacheConfig(method="snapkv", snap_budget=budget, snap_obs_window=8, snap_n_sink=2, seed=42)
    caches = KVCacheBuilder.for_model(model, cfg)
    mx.random.seed(0)
    out = "".join(r.text for r in stream_generate(model, tokenizer, prompt, max_tokens=20, prompt_cache=caches))
    print(budget, repr(out))
```

Output (temperature=0, deterministic):

| budget | tokens evicted | output |
|---|---|---|
| 200 | 0 (no eviction) | "The history and evolution of computer operating systems have been a fascinating journey..." (coherent) |
| 60 | 5 | "The history and evolution of computer operating systems can be traced back..." (coherent) |
| 50 | 15 | "Sure, I can help you with the six-paragraph essay about the history..." (drifting off-topic) |
| 40 | 25 | "Great to meet you, @aimee. I'm glad to meet you too, @a" (unrelated) |
| 35 | 30 | " 1. 贴纸\n  2. 带屏\n  " (raw Chinese-token garbage) |

## Root cause

`Attention.__call__` (mlx_lm's `qwen2.py`, representative of all models) does, per layer:

```python
queries = self.rope(queries, offset=cache.offset)
keys = self.rope(keys, offset=cache.offset)
keys, values = cache.update_and_fetch(keys, values)      # <- SnapKV evicts HERE
output = scaled_dot_product_attention(queries, keys, values, cache=cache, scale=self.scale, mask=mask)
```

`mask` was already built *before* `update_and_fetch` ran, via `create_attention_mask(h, cache=cache)` → `cache.make_mask(N, ...)`. `SnapKVKVCache` doesn't override `make_mask`, so it inherits `KVCache.make_mask`, which returns the *string* `"causal"` when `offset == 0` (true for the very first prefill call). `mx.fast.scaled_dot_product_attention` accepts `mask="causal"` with mismatched query/key lengths **without raising** — verified directly:

```python
mx.fast.scaled_dot_product_attention(q, k, v, scale=1.0, mask="causal")  # q: (1,2,65,64), k/v: (1,2,40,64) -- no error, wrong output
```

but its internal causal-masking logic assumes the `k` axis represents the last `len(k)` positions ending at each query's own position (the standard incremental-decode shape contract) — never true here, since SnapKV's kept keys are a *sparse, reordered subset* of the original 65 positions (whichever scored highest by observation-window attention), not a contiguous trailing window. Every query in the prefill step ends up attending to the wrong keys relative to its true position.

## Scope

- Only affects **prefill** eviction (the first `update_and_fetch` call, and every later `_process_prefill_chunk` call for chunked prefill) — decode-time eviction isn't applicable since SnapKV never evicts decode tokens.
- Only manifests once eviction actually drops rows (`budget < prompt_length`), i.e. exactly the case the feature exists for.
- This bug was **completely invisible in production** until now: `mlx_lm.server`'s batching path (`hasattr(cache, "merge")`) silently substituted a plain `BatchKVCache` for every `snapkv` request (see VeloxQuant-MLX#358 / VeloxQuant-Studio#27's now-merged fix), which never evicts anything — so `SnapKVKVCache.update_and_fetch` never actually ran with an evicted key set on a served request until #358's fix forced the real `_serve_single` path.

## Suggested fix directions (not yet attempted — architectural, needs its own PR)

1. **Explicit mask using `kept_indices`.** `snapkv_compress`/`_snapkv_compress_batched` already computes `kept_indices` (ascending original positions) internally but only the `(2)` `(k, v)` values are returned to `update_and_fetch`. Plumb `kept_indices` through, and override `make_mask` (or otherwise force an explicit array mask, never the `"causal"` string shortcut, whenever eviction has occurred) so each query position `i` only attends to kept key `j` when `kept_indices[j] <= i`.
2. **Defer eviction until after this step's own attention.** Let the first prefill step compute attention over the *full* un-evicted key set (correct, standard causal behavior for the tokens' own turn), and only compress the *stored* cache afterward for future steps to build on — i.e. eviction should never shrink the keys handed back to the very attention call whose queries just produced them.

Either fix needs careful, real live verification (not just unit tests) given how easily this regresses silently — the existing `test_offset_tracks_true_position_not_retained_rows` regression test does not catch it because it never checks attention output correctness, only offset bookkeeping.

## Environment

Reproduced with `mlx-community/Qwen2.5-0.5B-Instruct-4bit`, `mlx_lm` version installed in the scratch checkout used for live verification (see VeloxQuant-Studio issue #27's other PRs for the standard verification protocol).

**Comment by rajveer43** [comment]
## Update: same bug reproduces in `SqueezeAttentionCache` (VeloxQuant-Studio issue #28)

While live-verifying issue #28 (squeeze), fixing squeeze's own `merge()`/field-leak bugs let its real per-request eviction path run for the first time (same mechanism as this issue's original discovery) and exposed the identical corruption.

`SqueezeAttentionCache` also inherits base `KVCache.make_mask` (no override), so the same `mask="causal"` string shortcut applies. Repro (same prompt/harness as above, `squeeze_strength=1.0`, `squeeze_n_sink=4`, `max_tokens=20`, temperature 0):

| budget | output |
|---|---|
| 4096 (no eviction) | "The history and evolution of computer operating systems can be traced back to the dawn of computing, where the" (coherent) |
| 60 | "- - - - - - - - - - - - - - - - - - - -" |
| 50 | "about Q: Q: Q: Q: Q: Q: Q: Q: Q: Q" |
| 40 | "\`\`\`python\n\`\`\`python\n\`\`\`python\n\`\`\`python..." |
| 32 | " 你好，我来为"你好，你 你好，你 你好，你 " |

Same coherence cliff shape as SnapKV's table above, confirming this is not SnapKV-specific.

### Scope check across the eviction family

All caches that inherit base `KVCache.make_mask` (i.e. don't override it) share this exposure. Checked which eviction-family caches define their own `make_mask`:

```
h2o_cache.py        -- inherits base make_mask
tova_cache.py        -- inherits base make_mask
pyramidkv_cache.py   -- inherits base make_mask
chunkkv_cache.py     -- inherits base make_mask
cam_cache.py         -- inherits base make_mask
snapkv_cache.py      -- inherits base make_mask
squeeze_cache.py     -- inherits base make_mask
```

None of the eviction-family caches override `make_mask`. This is very likely a systemic bug across the whole family (H2O, TOVA, PyramidKV, ChunkKV, CaM, SnapKV, SqueezeAttention) wherever prefill eviction actually triggers within a single step — not just SnapKV/SqueezeAttention, those are just the two verified live so far via the #358 merge-fix rollout exposing real eviction on served requests for the first time. Have not reproduced the other four (h2o/tova/pyramidkv/chunkkv/cam) directly yet — flagging so a fix here is scoped to the shared base-class masking logic (or a shared mixin), not patched per-cache, since a per-cache patch would need to be repeated N more times as each method's own `method-verification` issue comes up and its merge() fix un-masks the same underlying bug.

Still deliberately not fixing this in #27's or #28's PRs, per the same scoping decision as before — this needs the `kept_indices`-based mask plumbing (or deferred eviction) discussed above, and ideally a single fix applied to whatever shared base/mixin these caches route through, rather than N separate patches.

**Comment by rajveer43** [comment]
## Confirmed a third time: streaming_llm (issue #29)

`StreamingLLMKVCache` reproduces this too, and arguably more directly than snapkv/squeeze — it evicts *continuously* (every `update_and_fetch` call, decode included, not just at prefill), yet still reports `self.offset = self._true_offset` (the true absolute position) while only `n_keep = stream_n_sink + stream_window_size` rows actually exist in `self.keys`/`self.values`. `make_mask` is inherited unchanged from base `KVCache`, so `create_attention_mask(..., offset=self.offset, ...)` is built against the true offset every single step, not just once at prefill — the query/key-length mismatch this issue describes is present on every decode step once the window saturates, not only during the initial prompt.

Found while live-verifying VeloxQuant-Studio issue #29, same discovery path as #27/#28: fixing the unrelated `merge()` batching-substitution bug (VeloxQuant-MLX#358) let real eviction run for the first time, exposing this.

### Repro (real `veloxquant serve`, `mlx-community/Qwen2.5-0.5B-Instruct-4bit`, temp=0)

Prompt: "Write a short haiku about the ocean, then explain what a haiku is in one sentence." (49 tokens), `stream_n_sink=4`, sweeping `stream_window_size`:

| `stream_window_size` | total window | output |
|---|---|---|
| 20 | 24 | coherent, on-topic |
| 8  | 12 | fluent but off-topic (drifts into an unrelated MCQ) |
| 4  | 8  | degenerate: `"Hello! How can I assist you today? - 你有什么建议吗？（ ）\nA. 1\nB. 1\nC. 1\nD. 100\n\`\`\`\n\n\`\`\`\n..."` |

Confirmed via `/v1/kv/stats` (after also fixing a separate, unrelated `tokens_kept` naming bug for streaming_llm in the #29 PR) that eviction was genuinely active at each of these settings — `tokens.retained` matched `n_layers * (n_sink + window_size)` exactly in each case, ruling out "the fix didn't take."

### Cache files still confirmed to inherit base `make_mask` (no override)

Unchanged from the #27/#28 findings — h2o, tova, pyramidkv, chunkkv, cam, snapkv, squeeze, and now streaming_llm all inherit it. This remains deliberately unfixed in the #29 PR, same scoping as #27/#28 — a real fix belongs in the shared base-class masking logic (or a shared mixin), not patched per-cache.
